perf/perf_fuzzer.py

```diff
@@ -0,0 +1,97 @@
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright: 2018 IBM
+# Author:Kamalesh Babulal <kamalesh@linux.vnet.ibm.com>
+#
+
+import platform
+import os
+from avocado import Test
+from avocado import main
+from avocado.utils import archive, build, process, distro, genio
+from avocado.utils.software_manager import SoftwareManager
+
+
+class Perffuzzer(Test):
+
+    """
+    This test run the perf fuzzer, which fuzz's the Linux perf_event
+    system call interface.
+    Source: https://github.com/deater/perf_event_tests/fuzzer
+    (http://web.eece.maine.edu/~vweaver/projects/perf_events/fuzzer/)
+    This test might crash the kernel
+    :avocado: tags=privileged
+    """
+
+    @staticmethod
+    def run_cmd_out(cmd):
+        return process.system_output(cmd, shell=True, ignore_status=True, sudo=True)
+
+    def setUp(self):
+        '''
+        Install the packages
+        '''
+        # Check for basic utilities
+        smm = SoftwareManager()
+        detected_distro = distro.detect()
+        deps = ['gcc', 'make']
+        if 'Ubuntu' in detected_distro.name:
+            kernel_ver = platform.uname()[2]
+            deps.extend(['linux-tools-common', 'linux-tools-%s'
+                         % kernel_ver])
+        # FIXME: "redhat" as the distro name for RHEL is deprecated
+        # on Avocado versions >= 50.0.  This is a temporary compatibility
+        # enabler for older runners, but should be removed soon
+        elif detected_distro.name in ['rhel', 'SuSE', 'fedora', 'redhat']:
+            deps.extend(['perf'])
+        else:
+            self.cancel("Install the package for perf supported by %s"
+                        % detected_distro.name)
+        for package in deps:
+            if not smm.check_installed(package) and not smm.install(package):
+                self.cancel('%s is needed for the test to be run' % package)
+
+    def build_perf_test(self):
+        """
+        Building the perf event test suite
+        """
+        tarball = self.fetch_asset('perf-event.zip', locations=[
+                                   'https://github.com/deater/'
+                                   'perf_event_tests/archive/'
+                                   'master.zip'], expire='7d')
+        archive.extract(tarball, self.workdir)
+        self.sourcedir = os.path.join(self.workdir, 'perf_event_tests-master')
+        if build.make(self.sourcedir, extra_args="-s -S") > 0:
+            self.fail("Building perf event test suite failed")
+
+    def execute_perf_fuzzer(self):
+        os.chdir(self.sourcedir)
+        genio.write_one_line("/proc/sys/kernel/perf_event_paranoid", "-1")
+        if "-1" not in genio.read_one_line("/proc/sys/kernel/perf_event_paranoid"):
+            self.error("Unable to set perf_event_paranoid to -1 ")
```

Comment:
i think `self.fail ` will be more suitable option here as it also part of test

Reply:
ah so it is system issue so it think it comes as `pre requisite` so better option is self.cancel

Reply:
I beg to differ, here we are writing into the `sys` interface and the kernel returns error. It should be considered error, rather than cancel. It really confusing on where to use `self.{fail,cancel, error}` do we have guidelines on it ?

Reply:
@kamalesh-babulal hey no hard feeling just a suggestion , i am suggesting here self.cancel as it seems me system pre requisite ,  you are owner of test if you feel if it is like self.error should be fine   as it just kind of suggestion as it just kind of representation of issues ,